File: JavaIntroNeuralNetworkEdition2/src/com/heatonresearch/book/introneuralnet/ch3/console/ConsoleHopfield.java
/**
 * Introduction to Neural Networks with Java, 2nd Edition
 * Copyright 2008 by Heaton Research, Inc. 
 * http://www.heatonresearch.com/books/java-neural-2/
 * 
 * ISBN13: 978-1-60439-008-7  	 
 * ISBN:   1-60439-008-5
 *   
 * This class is released under the:
 * GNU Lesser General Public License (LGPL)
 * http://www.gnu.org/copyleft/lesser.html
 */
package com.heatonresearch.book.introneuralnet.ch3.console;

import com.heatonresearch.book.introneuralnet.neural.hopfield.HopfieldNetwork;

/**
 * Chapter 3: Using a Hopfield Neural Network
 * 
 * ConsoleHopfield: Simple console application that shows how to
 * use a Hopfield Neural Network.
 * 
 * @author Jeff Heaton
 * @version 2.1
 */
public class ConsoleHopfield {

	/**
	 * Convert a boolean array to the form [T,T,F,F]
	 * 
	 * @param b
	 *            A boolen array.
	 * @return The boolen array in string form.
	 */
	public static String formatBoolean(final boolean b[]) {
		final StringBuilder result = new StringBuilder();
		result.append('[');
		for (int i = 0; i < b.length; i++) {
			if (b[i]) {
				result.append("T");
			} else {
				result.append("F");
			}
			if (i != b.length - 1) {
				result.append(",");
			}
		}
		result.append(']');
		return (result.toString());
	}

	/**
	 * A simple main method to test the Hopfield neural network.
	 * 
	 * @param args
	 *            Not used.
	 */
	public static void main(final String args[]) {

		// Create the neural network.
		final HopfieldNetwork network = new HopfieldNetwork(4);
		// This pattern will be trained
		final boolean[] pattern1 = { true, true, false, false };
		// This pattern will be presented
		final boolean[] pattern2 = { true, false, false, false };
		boolean[] result;

		// train the neural network with pattern1
		System.out.println("Training Hopfield network with: "
				+ formatBoolean(pattern1));
		network.train(pattern1);
		// present pattern1 and see it recognized
		result = network.present(pattern1);
		System.out.println("Presenting pattern:" + formatBoolean(pattern1)
				+ ", and got " + formatBoolean(result));
		// Present pattern2, which is similar to pattern 1. Pattern 1
		// should be recalled.
		result = network.present(pattern2);
		System.out.println("Presenting pattern:" + formatBoolean(pattern2)
				+ ", and got " + formatBoolean(result));

	}

}

File: CSIntroNeuralNetworkEdition2/Chapter03Console/ConsoleHopfield.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using HeatonResearchNeural.Hopfield;

namespace Chapter03Console
{

    /// <summary>
    /// Chapter 3: Using a Hopfield Neural Network
    ///
    /// ConsoleHopfield: Simple console application that shows how to
    /// use a Hopfield Neural Network.
    /// </summary>
    public class ConsoleHopfield
    {
        /// <summary>
        /// Convert a boolean array to the form [T,T,F,F]
        /// </summary>
        /// <param name="b">A boolen array.</param>
        /// <returns>The boolen array in string form.</returns>
        public static String FormatBoolean(bool[] b)
        {
            StringBuilder result = new StringBuilder();
            result.Append('[');
            for (int i = 0; i < b.Length; i++)
            {
                if (b[i])
                {
                    result.Append("T");
                }
                else
                {
                    result.Append("F");
                }
                if (i != b.Length - 1)
                {
                    result.Append(",");
                }
            }
            result.Append(']');
            return (result.ToString());
        }

        /**
         * A simple main method to test the Hopfield neural network.
         * 
         * @param args
         *            Not used.
         */
        static void Main(string[] args)
        {

            // Create the neural network.
            HopfieldNetwork network = new HopfieldNetwork(4);
            // This pattern will be trained
            bool[] pattern1 = { true, true, false, false };
            // This pattern will be presented
            bool[] pattern2 = { true, false, false, false };
            bool[] result;

            // train the neural network with pattern1        
            Console.WriteLine("Training Hopfield network with: "
                    + FormatBoolean(pattern1));
            network.Train(pattern1);
            // present pattern1 and see it recognized
            result = network.Present(pattern1);
            Console.WriteLine("Presenting pattern:" + FormatBoolean(pattern1)
                    + ", and got " + FormatBoolean(result));
            // Present pattern2, which is similar to pattern 1. Pattern 1
            // should be recalled.
            result = network.Present(pattern2);
            Console.WriteLine("Presenting pattern:" + FormatBoolean(pattern2)
                    + ", and got " + FormatBoolean(result));

        }
    }
}
